### ISSUE TITLE
async.race = execute in parallel, but only use the first result

### ISSUE THREAD

**Issue by rprieto** [caolan/async#568]
I found myself needing this type of function a few times:

- execute several async operations in parallel
- call `done` whenever the first one finished, and ignore any subsequent results

It's easily implemented as follows, but does it belong in `async`? Happy to submit a PR with unit tests.

```js
async.first = function(ops, done) {
  var i = 0;
  var response = false;
  for (i = 0; i < ops.length; ++i) {
    ops[i](function(err, data) {
      if (!response) {
        response = true;
        done(err, data);
      }
    });
  }
};
```

**Comment by kumavis** [comment]
+1 do want

**Comment by kumavis** [comment]
an alternate mode would be to ignore errors, and keep waiting for the first non-failure response

**Comment by ORESoftware** [comment]
+1

**Comment by ORESoftware** [comment]
@rprieto  you might need to create some logic to check the result of the request. for example, if the request returns an error, we might not want the first one, or if the first request doesn't have the data you want, you might not want the first result either. So there might need to be some conditional check in there somewhere.

So for generic usage, we want not just the first result but the first result that meets some condition. So your version of async.first might want  to take one more callback function that contains some conditional logic that returns a boolean.

**Comment by ORESoftware** [comment]
something that looks like:

```javascript
async.first = function(ops, func, done) {  // developer defined func to check condition
  var length = ops.length;
  var responses = 0;
  var success = false;
  for (var i = 0; i < length; i++) {
    ops[i](function(err, data) {
      responses++;
      if (!success) {
        if(func(err,data)){  
        success = true;
        done(err, data);
       }
     else if(responses >= length){  //we have received all responses, but none met the condition
          success = true;   // probably should rename success var, but ok for now
          done(new Error('all responses received, but no responses met condition', data);
   }
 else{
   //do nothing, wait for the next response
  }
      }
    });
  }
};
```

**Comment by ORESoftware** [comment]
so that might look like this in action:

```javascript
async.first([], function(err,data){
    if(err){
          return false;
      }
      else if(data.error){
         return false;
     }
      else{
         return true;
    }
}, function done(err, data){
      // do your thing
});
```

**Comment by aearly** [comment]
I think `async.race` is a good name for this -- it has some symmetry with [Promise.race](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race)

**Comment by ORESoftware** [comment]
I thought about that name too (or maybe some other mention of race seeded my thought process)

**Comment by Moeriki** [comment]
I implemented an asyncOneOf / asyncOneOfParallel because I liked this as well.

https://gist.github.com/Moeriki/4d5053052d1302f08d98

**Comment by aearly** [comment]
Closed by #1038 !

**Comment by kumavis** [comment]
:smile_cat: :clap:
